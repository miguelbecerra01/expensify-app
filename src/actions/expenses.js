//https://www.npmjs.com/package/uuid
import uuid from 'uuid';
import database from '../firebase/firebase';

//Actions 

//steps-> Syncronous calls
//component calls action generator
//action generator returns object
//component dispatches object
//redux store changes

//steps -> Asyncronous calls (save to firebase)
//component calls action generator
//action generator returns function
//component dispatches function (?) using a redux-module redux-thunk
//(a thunk is a subroutine used to inject an additional calculation into another subroutine)
//function runs (has the ability to dispatch other actions and do whatever it wants)


// ADD_EXPENSE
// export const addExpense = (
//     { description = '',
//         note = '',
//         amount = 0,
//         createdAt = 0
//     } = {}
// ) => ({
//     type: 'ADD_EXPENSE',
//     expense: {
//         id: uuid(),
//         description,
//         note,
//         amount,
//         createdAt
//     }
// });

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});


export const startAddExpense = (expenseData = {}) => {
    //will work only because we added the redux middleware using redux-thunk
    return (dispatch) => {
        //extract and create the parameters from the object expenseData -> destructuring
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;

        const expense = { description, note, amount, createdAt };

        //with return we add the hability to make promises chaining
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};


// REMOVE_EXPENSE 
//with parentesis after the arrow function means that returns the object
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

