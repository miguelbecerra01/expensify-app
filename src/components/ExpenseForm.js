import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

//its a class because we need to use component state
export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: false,
            submitType: props.submitType === 'Add' ? 'Add' : 'Edit'
        };
    }
    // state = {
    //     description: '',
    //     note: '',
    //     amount: '',
    //     createdAt: moment(),
    //     calendarFocused: false,
    //     error: false
    // };
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => {
            this.state.error = !description;
            return {
                description
            }
        });
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => {
                this.state.error = !amount;
                return {
                    amount
                };
            });
        }
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount.' }));
        } else {
            this.setState(() => ({ error: '' }));
            //call to the parent method in AddExpensePage, returning the value
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                //valueOf is from momentjs, because createdAt is a moment object
                //with valueOf returns in miliseconds 
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };
    render() {
        return (
            <div>

                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        name="description"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange} />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange} />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        firstDayOfWeek={1}
                        small={true}
                        isOutsideRange={() => false} />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        onChange={this.onNoteChange}
                        value={this.state.note}
                    >

                    </textarea>
                    <button >{this.state.submitType} Expense</button>
                </form>
            </div>
        );
    }
}