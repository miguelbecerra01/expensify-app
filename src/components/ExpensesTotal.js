import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';

import getExpensesTotal from '../selectors/expenses-total';

export const ExpensesTotal = (props) => (
    <div>
        {props.summary.quantity > 0 &&
            <p>Listing {props.summary.quantity} expense{props.summary.quantity > 1 && 's'}
                with a total of
            <span> {numeral(props.summary.total / 100).format('$0,0.00')}</span></p>}

    </div>
);

const mapStateToProps = (state) => {
    return {
        summary: getExpensesTotal(state.expenses)
    };
};

export default connect(mapStateToProps)(ExpensesTotal);
