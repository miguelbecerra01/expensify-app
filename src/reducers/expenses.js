//Expenses Reducer
const expensesReducerDefaultState = [];
const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            //return state.concat(action.expense);
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            //we need to destructure the array of objects in order to filter by id, otherwise it wont work    
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPENSE':
            //loop througth every item on the array
            return state.map((expense) => {
                if (expense.id === action.id) {
                    //spread operator for override the object with new values
                    //const expenseEdited = {...expense,...action.updates};
                    //return expenseEdited;
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense;
                }
            });
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
};


export default expensesReducer;