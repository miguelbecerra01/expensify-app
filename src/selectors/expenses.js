//Get visible expenses

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {

    return expenses.filter((expense) => {
        const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
        const textMatch = (expense.description.toLowerCase()).includes(text.toLowerCase());

        //if is a match
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            //sort descending
            return a.createdAt < b.createdAt ? 1 : -1;
        }
            //sort descending
        if (sortBy === 'amount') {
            //if is true b comes first if is false a comes first
            return a.amount < b.amount ? 1 : -1;
        }
    });
};

export default getVisibleExpenses;