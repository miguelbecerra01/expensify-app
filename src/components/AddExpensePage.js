import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <h1>Add Expense Page</h1>
                <ExpenseForm submitType={'Add'} onSubmit={this.onSubmit} />
            </div>
        );
    }
}


//https://react-redux.js.org/using-react-redux/connect-mapdispatch#connect-dispatching-actions-with-mapdispatchtoprops
//Its a way to return your dispatch functions allowing you to abstract them 
//away of the component itseft
const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
