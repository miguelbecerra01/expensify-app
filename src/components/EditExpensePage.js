import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';

const EditExpensePage = (props) => {
    return (
        <div>
            <h1>Edit Expense</h1>
            <ExpenseForm />
        </div>
    )
};

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => (expense.id === props.match.params.id))
    };
};

export default connect(mapStateToProps)(EditExpensePage);