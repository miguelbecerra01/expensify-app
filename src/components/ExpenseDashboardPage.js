import React from 'react';
import ReactDOM from 'react-dom';

import * as moment from 'moment/moment';
moment.locale('es');

import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesTotal from './ExpensesTotal';

const ExpenseDashboardPage = () => {

    return (
        <div>
            <hr></hr>
            <ExpensesTotal />
            <ExpenseListFilters />
            <ExpenseList />
        </div>
    );
}


export default ExpenseDashboardPage;