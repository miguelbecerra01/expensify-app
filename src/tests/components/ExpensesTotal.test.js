//1.- Import react, shallow, the component, and fixtures data

import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesTotal } from '../../components/ExpensesTotal';
import expenses from '../fixtures/expenses';

test('should render ExpensesTotal if no expenses', () => {
    const summary = {
        total: 0,
        quantity: 0
    };
    const wrapper = shallow(<ExpensesTotal summary={summary} />);
    expect(wrapper).toMatchSnapshot();

});

test('should render ExpensesTotal summary expenses', () => {
    const summary = {
        total: 10000,
        quantity: 1
    };
    const wrapper = shallow(<ExpensesTotal summary={summary} />);
    expect(wrapper).toMatchSnapshot();
});
