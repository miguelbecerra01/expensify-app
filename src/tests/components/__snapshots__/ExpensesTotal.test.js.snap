// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesTotal if no expenses 1`] = `<div />`;

exports[`should render ExpensesTotal summary expenses 1`] = `
<div>
  <p>
    Listing 
    1
     expense
    with a total of
    <span>
       
      $100.00
    </span>
  </p>
</div>
`;
